src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp

```diff
@@ -516,6 +559,508 @@ void mergeParallelMatMuls(const std::shared_ptr<ov::Model>& m, Context& ctx) {
+// gather with a host-side op. Lower vocab tensor to f16.
+// * - This case normally happens as a result of other
+// * - DictGather-related transformations
+HostGather::HostGather(Context::Ref ctx) {
+    auto pids = opp::wrap_type<ov::op::v0::Parameter>();
+    auto cvtids = opp::wrap_type<ov::op::v0::Convert>({pids});
+
+    auto qweight = opp::wrap_type<ov::op::v0::Parameter>();
+    auto qgthrw = opp::wrap_type<ov::op::v8::Gather>({qweight, cvtids, opp::any_input()});
+
+    auto callback = [=](ov::pass::pattern::Matcher& m) {
+        auto& node_to_output = m.get_pattern_value_map();
+        auto out_shape = node_to_output.at(qgthrw).get_shape();
+        auto matched_out_qweight = node_to_output.at(qweight);
+        auto qweight_type = matched_out_qweight.get_element_type();
+
+        auto matched_out_gather = node_to_output.at(qgthrw);
+
+        auto sole_reader = [](ov::Output<ov::Node> out) {
+            const auto readers = out.get_target_inputs();
+            NPUW_ASSERT(readers.size() >= 1);
+            return readers.begin()->get_node();
+        };
+
+        if (out_shape.back() >= 2048 && (qweight_type == ov::element::f16 || qweight_type == ov::element::f32) &&
+            (matched_out_gather.get_target_inputs().size() > 1 ||
+             ov::is_type<ov::op::v0::Convert>(sole_reader(matched_out_gather)))) {
+            auto matched_node_qweight = node_to_output.at(qweight).get_node_shared_ptr();
+            auto matched_node_ids = node_to_output.at(pids).get_node_shared_ptr();
+            auto matched_out_gthr = node_to_output.at(qgthrw);
+            auto matched_qweight = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_qweight);
+            auto matched_ids = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_ids);
+
+            if (qweight_type == ov::element::f32) {
+                ctx.get().to_f16(matched_qweight);
+            }
+            auto new_param = ctx.get().host_gather(matched_qweight, matched_ids);
+            std::shared_ptr<ov::Node> new_cvt;
+            if (qweight_type == ov::element::f16) {
+                new_cvt = new_param;
+            } else {
+                new_cvt = std::make_shared<ov::op::v0::Convert>(new_param, ov::element::f32);
+            }
+            NPUW_ASSERT(new_cvt);
+            for (auto&& r : matched_out_gthr.get_target_inputs()) {
+                r.replace_source_output(new_cvt);
+            }
+            return true;  // Root has changed
+        }
+        return false;  // Root hasn't changed (yet)
+    };
+    register_matcher(std::make_shared<opp::Matcher>(qgthrw, "HostGather"), std::move(callback));
+}
+
+// Identify the case* where the gather is applied on a compressed
+// (symmetric) vocab tensor. Both CW and GQ paths are supported.
+//
+// FIXME: This may be inefficient: 4x-es the memory consumption
+// due to i4-to-fp16 conversion.
+HostGatherDQ::HostGatherDQ(Context::Ref ctx) {
```

Comment:
It would be RAM and not NPU memory, right?

Reply:
it all goes to L0 memory.